shimming-toolbox/shimmingtoolbox/files/file.py

```diff
@@ -0,0 +1,566 @@
+                    raise e
+                return default_value
+        return wrapper
+    return decorator
+
+NIFTI_EXTENSIONS = ('.nii.gz', '.nii')
+DEFAULT_SUFFIX = '_saved.nii.gz'
+
+class NiftiFile:
+    def __init__(self, fname_nii: str, path_output: str = None, json_needed: bool = True) -> None:
+        if not isinstance(fname_nii, str):
+            raise TypeError("fname_nii must be a string")
+        if not any(fname_nii.endswith(ext) for ext in NIFTI_EXTENSIONS):
+            raise ValueError(f"File must end with one of {NIFTI_EXTENSIONS}")
+        
+        # Convert relative path to absolute path
+        self.fname_nii: str = os.path.abspath(fname_nii)
+        self.path_nii: str = self.get_path_nii()
+        self.nii: nib.Nifti1Image
+        self.data: np.ndarray
+        self.nii, self.data = self.load_nii()
+        self.filename: str = self.get_filename()
+        self.json: dict | None = self.load_json(json_needed)
+        self.header = self.nii.header
+        self.affine = self.nii.affine
+        self.shape = self.data.shape
+        self.ndim = self.data.ndim
+        self.path_output = path_output if path_output else os.getcwd()
+        
+    def __eq__(self, other: nib.Nifti1Image) -> None:
+        """Override the = operator to set NiftiFile data from a nibabel image.
+        
+        Args:
+            other (nib.Nifti1Image): The NiBabel image to set
+            
+        Raises:
+            TypeError: If other is not a nibabel.Nifti1Image
+            
+        Returns:
+            NiftiFile: Self for method chaining
+        """
+        self.set_nii(other)
+        return self  # Return self for method chaining
+    
+    def load_nii(self):
+        """ Load a NIfTI file and return the NIfTI object and its data.
+        Args:
+            fname_nii (str): Path to the NIfTI file.
+
+        Raises:
+            ValueError: If the provided path does not exist or is not a valid NIfTI file.
+            
+        Returns:
+            nib.Nifti1Image: The loaded NIfTI image object.
+            numpy.ndarray: The data contained in the NIfTI file.
+        """
+        if not os.path.exists(self.fname_nii):
+            raise ValueError("Not an existing NIFTI path")
+        nii = nib.load(self.fname_nii)
+        data = nii.get_fdata()
+        
+        return nii, data
+    
+    def load_json(self, json_needed: bool = True) -> dict | None:
+        """ Load the JSON file corresponding to the NIfTI file.
+        The JSON file is expected to be in the same directory as the NIfTI file
+        and have the same base name.
+
+        Args:
+            None
+
+        Returns:
+            dict: The content of the JSON file if found, otherwise None.
+        """
+        json_path = self.get_json(json_needed)
+        if json_path is not None:
+            with open(json_path, 'r') as f:
+                return json.load(f)
+        else:
+            return None
+        
+    def save(self) -> None:
```

Comment:
Save JSON sidecar as well?

Reply:
The JSON will likely not have changed, put if we save the NIfTI somewhere else, it won't have its associated JSON sidecar. The point of the class is to make it easier to have sidecars with their .nii files. When we save the NIfTI, I think we should save the JSON with it.

Reply:
Added in [efca1e8](https://github.com/shimming-toolbox/shimming-toolbox/pull/614/commits/efca1e8b46480aa218845df5bac23d179b6e8e6b)